web/releases/tenzir-node-v5.1/index.md

```diff
@@ -0,0 +1,183 @@
+---
+title: "Tenzir Node v5.1: Moving Fields"
+slug: tenzir-node-v5.1
+authors: [dominiklohmann]
+date: 2025-04-25
+tags: [release, node]
+comments: true
+---
+
+[Tenzir Node v5.1][github-release] enhances TQL with ways to access and move fields, bitwise functions, and a native integration with Google Cloud Logging.
+
+![Tenzir Node v5.1](tenzir-node-v5.1.svg)
+
+[github-release]: https://github.com/tenzir/tenzir/releases/tag/v5.1.0
+
+<!-- truncate -->
+
+## Trailing Question Mark: The Preferred Way to Suppress Warnings
+
+With Tenzir Node v5.1, we're introducing a more elegant way to access fields
+without triggering warnings. While the `.?field` operator was useful, the new
+trailing question mark syntax (`field?`) is more intuitive and works in more
+contexts:
+
+```tql
+from {
+  record: {nested: 42},
+  empty_record: {}
+}
+value1 = record.nested?       // 42, no warning
+value2 = empty_record.nested? // null, no warning
+array_item = [1, 2, 3][5]?    // null, no warning for out-of-bounds
+```
+
+This new syntax is not only more readable for top-level fields (no need for
+`this.?field`), but also extends to array indexing, making it a versatile
+addition to TQL. The older `.?` operator is now deprecated and will be removed
+in a future release.
+
+## Moving Fields with the `move` Keyword
+
+The new `move` keyword provides an elegant way to utilize fields and
+automatically drop them after use. This reduces boilerplate code and prevents
+accidentally leaving unused fields in your data.
+
+Consider this example:
+
+```tql
+from {xs: [1, 2, 3]}
+sum = sum(xs)
+drop xs
+```
+
+```tql
+{sum: 6}
+```
+
+Now, you can specify that the field should be moved within the assignment:
+
+```tql
+from {xs: [1, 2, 3]}
+sum = sum(move xs)
+```
+
+```tql
+{sum: 6}
+```
+
+Like other field accesses, `move` now also supports the trailing question mark
+syntax to suppress warnings for non-existent fields:
+
+```tql
+from {foo: 1}
+x = move foo?
+y = move bar?
+```
+
+```tql
+{
+  x: 1,
+  y: null,
+}
+```
+
+This is particularly useful when mapping events to a standardized schema where
+you want to keep track of fields you did not map yet:
+
+```tql
+from {unmapped: {…}}
+mapped = {
+  time: move unmapped.timestamp,
+  message: move unmapped.message
+}
+this = {
+  ...mapped,
+  // Keep all fields that weren't mapped as NDJSON.
+  unmapped: unmapped.print_ndjson(),
+}
+```
+
+## Bitwise Operations
+
+TQL now includes a comprehensive set of bitwise functions for numeric data
+processing:
+
+```tql
+from {
+  band: bit_and(5, 3),
+  bor: bit_or(5, 3),
+  bxor: bit_xor(5, 3),
+  bnot: bit_not(5),
+  shl: shift_left(5, 2),
+  shr: shift_right(5, 1),
+}
+```
+
+```tql
+{
+  band: 1,  // (0101 & 0011 = 0001)
+  bor: 7,   // (0101 | 0011 = 0111)
+  bxor: 6,  // (0101 ^ 0011 = 0110)
+  bnot: -6, // (~0101 = 1010)
```

Comment:
This assumes 4 bits integers.

Reply:
I added a sentence below explaining that integers are 64 bit and we're only showing the LSB for simplicity.